x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/LoadAuthorizedIndicesTimeChecker.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authz;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.settings.SettingsException;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xpack.core.security.authz.AuthorizationEngine;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+/**
+ * Checks the time spent evaluating authorized indices for a request.
+ * Has configurable logging levels based on execution time.
+ */
+class LoadAuthorizedIndicesTimeChecker implements Consumer<Collection<String>> {
+
+    private final Logger logger;
+    private final long startNanos;
+    private final AuthorizationEngine.RequestInfo requestInfo;
+    private final Thresholds thresholds;
+
+    LoadAuthorizedIndicesTimeChecker(Logger logger, long startNanos, AuthorizationEngine.RequestInfo requestInfo, Thresholds thresholds) {
+        this.logger = logger;
+        this.startNanos = startNanos;
+        this.requestInfo = requestInfo;
+        this.thresholds = thresholds;
+    }
+
+    @Override
+    public void accept(Collection<String> indices) {
+        done(indices);
+    }
+
+    private void done(Collection<String> indices) {
+        final long end = System.nanoTime();
+        final long millis = TimeUnit.NANOSECONDS.toMillis(end - startNanos);
+        final Level level = thresholds.getLogLevel(millis);
+        if (level == Level.WARN) {
+            logger.warn(
+                "Resolving [{}] indices for action [{}] and user [{}] took [{}ms] which is greater than the threshold of {}ms;"
+                    + " The index privileges for this user may be too complex for this cluster.",
+                indices.size(),
+                requestInfo.getAction(),
+                requestInfo.getAuthentication().getUser().principal(),
+                millis,
+                thresholds.warnThresholdMs
+            );
+        } else {
+            logger.log(
+                level,
+                "Took [{}ms] to resolve [{}] indices for action [{}] and user [{}]",
+                millis,
+                indices.size(),
+                requestInfo.getAction(),
+                requestInfo.getAuthentication().getUser().principal()
+            );
+        }
+    }
+
+    static final Setting<Boolean> LOGGING_ENABLED_SETTING = Setting.boolSetting(
+        "xpack.security.authz.timer.enabled",
+        true,
+        Setting.Property.NodeScope
```

Comment:
I wonder whether this should default to `false` and also be a dynamic setting. These log messages have generate quite a bit noise in the field and they are not always very useful. The performance improvements in 7.16 have made the slow loading scenarios reasonably predicable. I'd personally prefer turning these messages off by default.

Reply:
Yeah, I think you're right. Let me rework it.